pandas/io/parsers.py

```diff
@@ -1669,6 +1669,16 @@ def _convert_to_ndarrays(self, dct, na_values, na_fvalues, verbose=False,
 
                 # type specified in dtype param
                 if cast_type and not is_dtype_equal(cvals, cast_type):
+                    try:
+                        if (is_bool_dtype(cast_type) and
+                                not is_categorical_dtype(cast_type)
+                                and set(values) - set(col_na_values)):
```

Comment:
Can't we check if `na_count` > 0?

Reply:
I tried but na_count is 0 in some of the test cases (it is set before the casting stage)

Reply:
Wouldn't `na_count` have been set from `self._infer_types` ?  Which test fails if you check `na_count` ?

Reply:
This happens in `self._infer_types`:
```
if try_num_bool:
    try:
        result = lib.maybe_convert_numeric(values, na_values, False)
        na_count = isna(result).sum()
    except Exception:
        result = values
        if values.dtype == np.object_:
            na_count = parsers.sanitize_objects(result,
                                                na_values, False)
```
`lib.maybe_convert_numeric` fails because at this stage, `values` is still a list of strings including 'true' and 'false', so `na_count = isna(result).sum()` is never executed, and `na_count` is not set for the rest of the function

Reply:
My bad, I didn't write the test case properly. Using na_count works